### PR TITLE
Implement createLeveragedPosition for e2e tests

### PR DESCRIPTION
**Pull request by CoreyMartin**

### CODE REVIEW

File: contracts/PositionToken.sol [hunk 1]
```diff
@@ -5,11 +5,12 @@ import "hardhat/console.sol";
 
 import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
 import "@openzeppelin/contracts/token/ERC721/extensions/ERC721Burnable.sol";
+import "@openzeppelin/contracts/token/ERC721/extensions/ERC721Enumerable.sol";
 import "@openzeppelin/contracts/access/Ownable.sol";
 import "@openzeppelin/contracts/utils/Counters.sol";
 import "@openzeppelin/contracts/access/AccessControl.sol";
 
-contract PositionToken is ERC721, ERC721Burnable, AccessControl {
+contract PositionToken is ERC721, ERC721Burnable, ERC721Enumerable, AccessControl {
```
Comment (CoreyMartin) [inline contracts/PositionToken.sol:13]: This enables getting the tokenId of a given address which is useful for the tests and possibly for offchain purposes

File: contracts/Coordinator.sol [hunk 2]
```diff
@@ -65,10 +65,11 @@ contract Coordinator is ICoordinator, ReentrancyGuard {
     function depositCollateralUnderNFT(
         uint256 _nftId,
         uint256 _amountInOUSD,
-        address _sharesOwner
+        address _from
     ) external override {
+        _ousd.transferFrom(_from, address(this), _amountInOUSD);
         /// Transfer collateral to vault, mint shares to shares owner
-        uint256 shares = _vault.deposit(_amountInOUSD, _sharesOwner);
+        uint256 shares = _vault.deposit(_amountInOUSD, address(this));
```
Comment (CoreyMartin) [inline contracts/Coordinator.sol:72]: As discussed in review yesterday, coordinator will be the shares owner

File: contracts/PositionToken.sol [hunk 3]
```diff
@@ -64,8 +65,16 @@ contract PositionToken is ERC721, ERC721Burnable, AccessControl {
...
     }
 
-    /* Override required by Solidity: */
-    function supportsInterface(bytes4 interfaceId) public view override(ERC721, AccessControl) returns (bool) {
+    /* Overrides required by Solidity: */
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256 tokenId
+    ) internal virtual override(ERC721, ERC721Enumerable) {
```
Comment (yotamdaniel) [inline contracts/PositionToken.sol:73]: We need to add non reentrancy here ([link](https://app.gitbook.com/o/B7Qi247Rg5Gidqy0rmTN/s/fcHzCkSayJoCWD74wFz1/infrastructure-reviews/20220420-apy-finance-and-zellic) to gitbook note from talk with auditors, scroll to "Re-entrance/Design
")

File: test/ContractTestContext.test.ts [hunk 4]
```diff
@@ -7,7 +7,7 @@ describe("ContractTestContext", function () {
     let contractTestContext: ContractTestContext;
 
     before(async function () {
-        helperResetNetwork(defaultBlockNumber);
+        await helperResetNetwork(defaultBlockNumber);
```
Comment (yotamdaniel) [inline test/ContractTestContext.test.ts:10]: nice catch

File: contracts/interfaces/ICoordinator.sol [hunk 5]
```diff
@@ -16,11 +16,11 @@ interface ICoordinator {
     ///
     /// @param _nftId the Archimedes ERC-721 token id
     /// @param _amountInOUSD the amount of OUSD sent to Archimedes
-    /// @param _sharesOwner who to send shares to
+    /// @param _from who to deposit funds from
     function depositCollateralUnderNFT(
         uint256 _nftId,
         uint256 _amountInOUSD,
-        address _sharesOwner
+        address _from
```
Comment (yotamdaniel) [inline contracts/interfaces/ICoordinator.sol:23]: no need for from here, see comment above

File: test/Coordinator.ts [hunk 6]
```diff
@@ -39,10 +39,8 @@ describe("Coordinator Test suit", function () {
         before(async function () {
             // transfer OUSD from user to coordinator address
             // (this will happen in leverage engine in full Archimedes flow)
-            await r.externalOUSD.connect(endUserSigner).transfer(coordinator.address, addr1CollateralAmount);
```
Comment (yotamdaniel) [inline test/Coordinator.ts:42]: transfer doesnt happen in LevEng at the moment ( but it should)

**Review by yotamdaniel — APPROVED** [review]
looks good, merge once tests are passing

**Pull request merged**